RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,509 @@
+    functions `empty` and `empty_like`?
+
+4.  The fill-value of a sparse tensor can be acquired via
+    `fill_value()` method that returns a strided `torch.Tensor`
+    instance that has the same dtype and storage location properties
+    as the elements of the sparse tensor:
+
+    - `A.fill_value().dtype == A.dtype`
+    - `A.fill_value().device == A.device`
+
+    The fill-value of a sparse non-hybrid tensor is a scalar tensor
+    `torch.tensor(0, dtype=A.dtype, device=A.device)`.
+
+5.  The fill-value of a hybrid sparse tensor has the same shape as the
+    dense part of the tensor.
+
+    For instance, for a sparse tensor `A` in COO storage format we
+    have:
+
+    - `A.fill_value().shape == A.values().shape[1:]`
+
+    While an obvious alternative to tensor fill-value would be a
+    scalar fill-value, the scalar fill-value would complicate the
+    implementation of algorithms for hybrid tensors. For
+    instance:
+
+    1. Accessing a dense element of a hybrid sparse tensor, the result
+       is a dense tensor. When this element corresponds to unspecified
+       element, one needs to construct a new tensor and fill it with
+       the scalar fill-value. A more efficient and simple way is to
+       just return the output of `fill_value()` that does would not
+       involve constructing a new tensor on each method call.
+
+    2. When applying a row-wise function (such as `softmax`, for
+       instance) to a hybrid tensor, each row may induce a different
+       fill-value.
+
+       For example, consider a 2D hybrid tensor with one sparse and
+       one dense dimension in COO format, and apply softmax along the
+       first dimension assuming the default fill-value 0:
+
+       ```python
+       >>> A = torch.sparse_coo_tensor(indices=[[0, 3]],
+                                       values=[[.11, .12], [.31, .32]],
+                                       size=(4, 2))
+       >>> A.to_dense()
+       tensor([[0.1100, 0.1200],
+               [0.0000, 0.0000],
+               [0.0000, 0.0000],
+               [0.3100, 0.3200]])
+       >>> A.to_dense().softmax(0)
+       tensor([[0.2492, 0.2503],
+               [0.2232, 0.2220],
+               [0.2232, 0.2220],
+               [0.3044, 0.3057]])
+       ```
+
+       Observe that the softmax result can be represented as a sparse
+       tensor with the same sparsity structure (same indices) as the
+       input only when the fill-value is a 1-D dense tensor, that is,
+
+       ```python
+       A.softmax(0) == torch.sparse_coo_tensor(indices=A.indices(),
+                                               values=[[0.2492, 0.2503], [0.3044, 0.3057]],
+                                               size=A.size(),
+                                               fill_value=[0.2232, 0.2220])
+       ```
+
+       If fill-value would be defined as scalar, the result of
+       `A.softmax(0)` would need to be a dense tensor or raise
+       exception about unsupported feature.
```

Comment:
`sparse.softmax` works today for this case right? And returns a dense tensor? I think that cannot be changed after release anymore.

Reply:
This again seems a trade-off: for new functions that gain sparse support and are `softmax`-like, you can return a sparse rather than a dense tensor for hybrid input (potentially saving memory) - but you're paying with increased memory use of the input tensor as well - even in use cases that never need the non-scalar fill value feature.

In addition, it makes implementation of algorithms more complex, and the interface as well (see, e.g., point 6 below which now needs a disclaimer about `.fill()` only working for non-hybrid tensors).

Reply:
> `sparse.softmax` works today for this case right? And returns a dense tensor? I think that cannot be changed after release anymore.

No. `torch.sparse.softmax` always returns a sparse tensor because the assumed fill-value is `-inf`.
`torch.softmax` does not support sparse tensors (because of the lack of fill-value).

Reply:
To be clear: "dense tensor" can use different storage format layouts. Typically it is `strided` but it could be also `sparse_coo` with `nnz==product(shape)`.

Reply:
> This again seems a trade-off: for new functions that gain sparse support and are `softmax`-like, you can return a sparse rather than a dense tensor for hybrid input (potentially saving memory) - but you're paying with increased memory use of the input tensor as well - even in use cases that never need the non-scalar fill value feature.

See https://github.com/pytorch/rfcs/pull/8#discussion_r487567061 that basically eliminates the trade-off concern.

Reply:
> In addition, it makes implementation of algorithms more complex, and the interface as well (see, e.g., point 6 below which now needs a disclaimer about `.fill()` only working for non-hybrid tensors).

I disagree. If the fill-value can be only a scalar, the result of `softmax` on sparse tensor is always dense (in the sense that all tensor elements are specified). The memory trade-off of this will be huge compared to the memory trade-off of having non-scalar fill-value for hybrid tensors.

The example in point 6. does not need the non-hybrid disclaimer, it would work also for hybrid tensors.

In general, when applying an operation to a tensor, the operation should result in a new tensor with the same storage format as the input tensor (unless the operation explicitly advertises or provides via `layout` or `out` argument a different behavior). This design pattern is used ,say, in Matlab, and IMHO this is a good pattern to follow.

Reply:
> In general, when applying an operation to a tensor, the operation should result in a new tensor with the same storage format as the input tensor (

That makes sense as a principle, agreed.

Reply:
I'll close this conversation as resolved as it seems to originate from a misunderstanding of hybrid tensors. Feel feel to reopen.